MantidQt/MantidWidgets/src/InstrumentView/PanelsSurface.cpp

```diff
@@ -231,9 +219,8 @@ void PanelsSurface::addFlatBank(ComponentID bankId,
     m_unwrappedDetectors.reserve(m_unwrappedDetectors.size() + nelem);
     for (int i = 0; i < nelem; ++i) {
       // setup detector info
-      Mantid::Geometry::IDetector_const_sptr det =
-          boost::dynamic_pointer_cast<const Mantid::Geometry::IDetector>(
-              assembly->getChild(i));
+      auto det = boost::dynamic_pointer_cast<const Mantid::Geometry::IDetector>(
+          assembly->getChild(i));
       if (!doneRotation) {
         pos0 = det->getPos();
```

Comment:
@OwenArnold I think @mantid-roman has a point that that would be too complicated. On the other hand the other `Infos` will probably not be available for many months.

However, I actually need a very similar functionality in my changes to `DetectorInfo` that are updating the positions of detectors when a parent is moved. Currently I am using `Instrument::getDetectorInBank`, and then obtain detectors IDs, which I then translate to indices. This is slow and painful.
Maybe there is a way to add `ComponentInfo` without any functionality for accessing components, but with a mechanism that provides easy and fast access to a list of all detector indices in a subtree? Basically what we want is

```cpp
std::vector<size_t> API::ComponentInfo::detectorIndices(const IComponent &comp) {
  // find and return all detector indices in subtree given by `comp`
}

for (const auto &detIndex : ws.componentInfo().detectorIndices(comp)) {
  ws.detectorInfo().position(detIndex);
}
```

This requires some more thought, in particular how we work with time indices in an elegant (i.e., ideally hidden) way.